src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1284 @@
+
+			processedRowCount += 1;
+		}
+
+		connectionList = ConnectionList(shardConnectionHash);
+
+		/* send binary footers to all shard placements */
+		resetStringInfo(copyOutState->fe_msgbuf);
+		AppendCopyBinaryFooters(copyOutState);
+		SendCopyDataToAll(copyOutState->fe_msgbuf, connectionList);
+
+		/* close the COPY input on all shard placements */
+		EndRemoteCopy(connectionList, true);
+
+		if (CopyTransactionManager == TRANSACTION_MANAGER_2PC)
+		{
+			PrepareRemoteTransactions(connectionList);
+		}
+
+		EndCopyFrom(copyState);
+		heap_close(distributedRelation, NoLock);
+
+		error_context_stack = errorCallback.previous;
+
+		/* check for cancellation one last time before committing */
+		CHECK_FOR_INTERRUPTS();
+	}
+	PG_CATCH();
+	{
+		List *abortConnectionList = NIL;
+
+		error_context_stack = errorCallback.previous;
+
+		/* roll back all transactions */
+		abortConnectionList = ConnectionList(shardConnectionHash);
+		EndRemoteCopy(abortConnectionList, false);
+		AbortRemoteTransactions(abortConnectionList);
+		CloseConnections(abortConnectionList);
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	/*
+	 * Ready to commit the transaction, this code is below the PG_TRY block because
+	 * we do not want any of the transactions rolled back if a failure occurs. Instead,
+	 * they should be rolled forward.
+	 */
+	CommitRemoteTransactions(connectionList);
+	CloseConnections(connectionList);
+
+	if (completionTag != NULL)
+	{
+		snprintf(completionTag, COMPLETION_TAG_BUFSIZE,
+				 "COPY " UINT64_FORMAT, processedRowCount);
+	}
+}
+
+
+/*
+ * LockAllShards takes shared locks on the metadata and the data of all shards in
+ * shardIntervalList. This prevents concurrent placement changes and concurrent
+ * DML statements that require an exclusive lock.
+ */
+static void
+LockAllShards(List *shardIntervalList)
+{
+	ListCell *shardIntervalCell = NULL;
+
+	/* lock shards in order of shard id to prevent deadlock */
+	shardIntervalList = SortList(shardIntervalList, CompareShardIntervalsById);
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		int64 shardId = shardInterval->shardId;
+
+		/* prevent concurrent changes to number of placements */
+		LockShardDistributionMetadata(shardId, ShareLock);
+
+		/* prevent concurrent update/delete statements */
+		LockShardResource(shardId, ShareLock);
+	}
+}
+
+
+/*
+ * CreateShardConnectionHash constructs a hash table used for shardId->Connection
+ * mapping.
+ */
+static HTAB *
+CreateShardConnectionHash(void)
+{
+	HTAB *shardConnectionsHash = NULL;
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(int64);
+	info.entrysize = sizeof(ShardConnections);
+	info.hash = tag_hash;
+
+	shardConnectionsHash = hash_create("Shard Connections Hash",
+									   INITIAL_CONNECTION_CACHE_SIZE, &info,
+									   HASH_ELEM | HASH_FUNCTION);
```

Comment:
Hm. I think we should specify HASH_CONTEXT here. We want the memory context to go away after an error.